Is TinyString and TinyLong needed?

Especially with `TinyString`, I find that the usage of the API become cumbersome:

```
auto cb = Vault::ConfigBuilder()
  .withDebug(true)
  .withHost(Vault::TinyString{host})
  .withTlsEnabled(true)
  .withPort(Vault::TinyString{"443"});
```

I'm not sure what `TinyString` actually brings to the table. I think it would be nicer if we can write:

```
auto cb = Vault::ConfigBuilder()
  .withDebug(true)
  .withHost(host)
  .withTlsEnabled(true)
  .withPort(443);
```

If we want to snowflake individual types for protection, the better way would be to have individual types for `Host`, `Token` etc.

```
struct Host {
  explicit Host(std::string v) : v_(std::move(v)) {}
  std::string v_;
};

...
auto cb = Vault::ConfigBuilder()
  .withDebug(true)
  .withHost(Host{host})
  .withTlsEnabled(true)
  .withPort(443);
```

I think that's a worthy goal. I strive to do the same myself.

As you can see, the typedefs do not create different types but rather type aliases to the same type. So in the case above, you can just as well write:

```cpp
auto cb = Vault::ConfigBuilder()
  .withDebug(true)
  .withHost(Vault::Port{host})
  .withTlsEnabled(true)
  .withPort(Vault::Host{"443"});
```

What you want is to create actual different types that have an underlying storage of type string.

Something like this would do it:
```cpp
struct Host {
  explicit Host(std::string v) : v_(std::move(v)) {}
  std::string v_;
};
```

However, if you want more shared functionality but still generate unique types, you can use a template like this.

`tinystring.cpp:`
```cpp
#include <string>

enum class Type {
  Host,
  Port
  // ...
};

template <Type T>
class TinyString {
public:
  TinyString() = default;
  virtual ~TinyString() {}

  TinyString(std::string value) : value_(std::move(value)) {}

  friend std::ostream &operator<<(std::ostream &os, const TinyString &object) { return os << object.value(); }
  friend std::string operator+(const std::string &string, const TinyString &tiny) { return string + tiny.value(); }
  friend std::string operator+(const TinyString &tiny, const std::string &string) { return tiny.value() + string; }
  friend std::string operator+(const char *string, const TinyString &tiny) { return string + tiny.value(); }
  friend std::string operator+(const TinyString &tiny, const char *string) { return tiny.value() + string; }
  friend std::string operator+(const TinyString &tiny, const TinyString &other) { return tiny.value() + tiny.value(); }

  [[nodiscard]] bool empty() const { return value_.empty(); }
  [[nodiscard]] const char *c_str() const { return value_.c_str(); }
  [[nodiscard]] const std::string &value() const { return value_; }

protected:
  std::string value_;
};

using Host = TinyString<Type::Host>;
using Port = TinyString<Type::Port>;
// ...

int main() {
  Host h1{"google.com"};
  Host h2{"yahoo.com"};
  h2 = h1; // Should be okay

  Port p{"8080"};
  p = h1; // Error! Different types.
}
```

The result of compiling this using C++17 with clang is the following:
```
> clang++ -std=c++17 tinystring.cpp
tinystring.cpp:42:5: error: no viable overloaded '='
  p = h1; // Error! Different types.
  ~ ^ ~~
tinystring.cpp:10:7: note: candidate function (the implicit copy assignment operator) not viable: no known conversion
      from 'TinyString<Type::Host aka 0>' to 'const TinyString<1>' for 1st argument
class TinyString {
      ^
1 error generated.
```

--- a/include/VaultClient.h
+++ b/include/VaultClient.h
@@ -54,23 +54,23 @@ namespace Vault {
     struct Name##Detail {};                \
     using Name = Tiny<Name##Detail, long>; \
 
-  TINY_STRING(SecretId);
-  TINY_STRING(HttpResponseBodyString);
-  TINY_STRING(Url);
-  TINY_STRING(Path);
-  TINY_STRING(Token);
-  TINY_STRING(Namespace);
-  TINY_STRING(RoleId);
-  TINY_STRING(Host);
-  TINY_STRING(Port);
-  TINY_STRING(Algorithm);
-  TINY_STRING(SecretMount);
-  TINY_STRING(RootCertificateType);
-  TINY_STRING(KeyType);
+  TINY_STRING(SecretId)
+  TINY_STRING(HttpResponseBodyString)
+  TINY_STRING(Url)
+  TINY_STRING(Path)
+  TINY_STRING(Token)
+  TINY_STRING(Namespace)
+  TINY_STRING(RoleId)
+  TINY_STRING(Host)
+  TINY_STRING(Port)
+  TINY_STRING(Algorithm)
+  TINY_STRING(SecretMount)
+  TINY_STRING(RootCertificateType)
+  TINY_STRING(KeyType)
 
-  TINY_LONG(HttpResponseStatusCode);
-  TINY_LONG(ConnectTimeout);
-  TINY_LONG(TTL);
+  TINY_LONG(HttpResponseStatusCode)
+  TINY_LONG(ConnectTimeout)
+  TINY_LONG(TTL)
 
   #undef TINY_STRING
   #undef TINY_LONG
@@ -128,7 +128,7 @@ namespace Vault {
     explicit HttpClient(Config &config);
 
     HttpClient(Config &config, HttpErrorCallback errorCallback);
-    virtual ~HttpClient() {}
+    virtual ~HttpClient() = default;
 
     [[nodiscard]] virtual std::optional<HttpResponse> get(const Url &url, const Token &token, const Namespace &ns) const;
     [[nodiscard]] virtual std::optional<HttpResponse> post(const Url &url, const Token &token, const Namespace &ns, const std::string& value) const;
@@ -201,7 +201,7 @@ namespace Vault {
 
   class AuthenticationStrategy {
   public:
-    virtual ~AuthenticationStrategy() {}
+    virtual ~AuthenticationStrategy() = default;
     virtual std::optional<AuthenticationResponse> authenticate(const Client &client) = 0;
   };
 
@@ -210,7 +210,7 @@ namespace Vault {
     Client(const Client &other, Token token);
     Client(Config &config, AuthenticationStrategy &authStrategy);
     Client(Config &config, AuthenticationStrategy &authStrategy, HttpErrorCallback httpErrorCallback);
-    virtual ~Client() {}
+    virtual ~Client() = default;
 
     [[nodiscard]] virtual bool is_authenticated() const { return !token_.empty(); }
     [[nodiscard]] virtual Url getUrl(const std::string &base, const Path &path) const;

--- a/include/VaultClient.h
+++ b/include/VaultClient.h
@@ -10,63 +10,67 @@
 #include <optional>
 #include <ostream>
 #include <string>
+#include <sstream>
 #include <utility>
 #include <vector>
 
 namespace Vault {
-  struct TinyString {
-    TinyString() = default;
+  template<typename Name, typename T>
+  struct Tiny {
+    Tiny() noexcept(std::is_nothrow_constructible_v<T>) = default;
+    explicit Tiny(const T &value) noexcept(std::is_nothrow_copy_constructible_v<T>) : value_(value) {}
+    explicit Tiny(T &&value) noexcept(std::is_nothrow_move_constructible_v<T>) : value_(std::move(value)) {}
 
-    explicit TinyString(std::string value) : value_(std::move(value)) {}
+    friend std::ostream &operator<<(std::ostream &os, const Tiny &object) { return os << object.value(); }
+    friend std::string operator+(const Tiny &tiny, const Tiny &other) { return tiny.value() + tiny.value(); }
 
-    friend std::ostream &operator<<(std::ostream &os, const TinyString &object) { return os << object.value(); }
-    friend std::string operator+(const std::string &string, const TinyString &tiny) { return string + tiny.value(); }
-    friend std::string operator+(const TinyString &tiny, const std::string &string) { return tiny.value() + string; }
-    friend std::string operator+(const char *string, const TinyString &tiny) { return string + tiny.value(); }
-    friend std::string operator+(const TinyString &tiny, const char *string) { return tiny.value() + string; }
-    friend std::string operator+(const TinyString &tiny, const TinyString &other) { return tiny.value() + tiny.value(); }
+    friend std::string operator+(const Tiny &tiny, const char *string) { return tiny.toString() + string; }
+    friend std::string operator+(const char *string, const Tiny &tiny) { return string + tiny.toString(); }
+    friend std::string operator+(const Tiny &tiny, const std::string &string) { return tiny.toString() + string; }
+    friend std::string operator+(const std::string &string, const Tiny &tiny) { return string + tiny.toString(); }
 
-    [[nodiscard]] bool empty() const { return value_.empty(); }
-    [[nodiscard]] const char *c_str() const { return value_.c_str(); }
-    [[nodiscard]] const std::string &value() const { return value_; }
+    explicit operator T& () const noexcept {
+      return value();
+    }
 
-  protected:
-    std::string value_;
-  };
+    [[nodiscard]] std::string toString() const {
+      std::ostringstream ss;
+      ss << value();
+      return ss.str();
+    }
 
-  typedef TinyString SecretId;
-  typedef TinyString HttpResponseBodyString;
-  typedef TinyString Url;
-  typedef TinyString Path;
-  typedef TinyString Token;
-  typedef TinyString Namespace;
-  typedef TinyString RoleId;
-  typedef TinyString Host;
-  typedef TinyString Port;
-  typedef TinyString Algorithm;
-  typedef TinyString SecretMount;
-  typedef TinyString RootCertificateType;
-  typedef TinyString KeyType;
-
-  struct TinyLong {
-    TinyLong() = default;
-    explicit TinyLong(const long value) : value_(value) {}
-
-    friend long operator+(const long addend, const TinyLong other) { return addend + other.value_; }
-    friend long operator+(const TinyLong other, const long addend) { return other.value_ + addend; }
-    friend long operator+(const TinyLong current, const TinyLong other) { return current.value_ + other.value_; }
-
-    friend std::string operator+(const char *other, const TinyLong current) { return other + std::to_string(current.value_); }
-
-    [[nodiscard]] long value() const { return value_; }
+    [[nodiscard]] bool empty() const noexcept { return value_.empty(); }
+    [[nodiscard]] const T &value() const noexcept { return value_; }
 
   protected:
-    long value_;
+    T value_;
   };
 
-  typedef TinyLong HttpResponseStatusCode;
-  typedef TinyLong ConnectTimeout;
-  typedef TinyLong TTL;
+  #define TINY_STRING(Name)                       \
+    struct Name##Detail {};                       \
+    using Name = Tiny<Name##Detail, std::string>; \
+
+  #define TINY_LONG(Name)                  \
+    struct Name##Detail {};                \
+    using Name = Tiny<Name##Detail, long>; \
+
+  TINY_STRING(SecretId);
+  TINY_STRING(HttpResponseBodyString);
+  TINY_STRING(Url);
+  TINY_STRING(Path);
+  TINY_STRING(Token);
+  TINY_STRING(Namespace);
+  TINY_STRING(RoleId);
+  TINY_STRING(Host);
+  TINY_STRING(Port);
+  TINY_STRING(Algorithm);
+  TINY_STRING(SecretMount);
+  TINY_STRING(RootCertificateType);
+  TINY_STRING(KeyType);
+
+  TINY_LONG(HttpResponseStatusCode);
+  TINY_LONG(ConnectTimeout);
+  TINY_LONG(TTL);
 
   namespace Algorithms {
     const static Vault::Algorithm SHA1 = Vault::Algorithm{"sha1"};

--- a/include/VaultClient.h
+++ b/include/VaultClient.h
@@ -72,6 +72,9 @@ namespace Vault {
   TINY_LONG(ConnectTimeout);
   TINY_LONG(TTL);
 
+  #undef TINY_STRING
+  #undef TINY_LONG
+
   namespace Algorithms {
     const static Vault::Algorithm SHA1 = Vault::Algorithm{"sha1"};
     const static Vault::Algorithm SHA2_224 = Vault::Algorithm{"sha2-224"};

--- a/src/support/HttpClient.cpp
+++ b/src/support/HttpClient.cpp
@@ -147,7 +147,7 @@ Vault::HttpClient::executeRequest(const Vault::Url& url,
 
     curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, connectTimeout_);
     curl_easy_setopt(curl, CURLOPT_HTTPHEADER, chunk);
-    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+    curl_easy_setopt(curl, CURLOPT_URL, url.toString().c_str());
     curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeCallback);
     curl_easy_setopt(curl, CURLOPT_WRITEDATA, &buffer);
 

--- a/src/support/HttpClient.cpp
+++ b/src/support/HttpClient.cpp
@@ -147,7 +147,7 @@ Vault::HttpClient::executeRequest(const Vault::Url& url,
 
     curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, connectTimeout_);
     curl_easy_setopt(curl, CURLOPT_HTTPHEADER, chunk);
-    curl_easy_setopt(curl, CURLOPT_URL, url.toString().c_str());
+    curl_easy_setopt(curl, CURLOPT_URL, url.value().c_str());
     curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeCallback);
     curl_easy_setopt(curl, CURLOPT_WRITEDATA, &buffer);
 

--- a/test/integration/system/SysTests.cpp
+++ b/test/integration/system/SysTests.cpp
@@ -27,7 +27,7 @@ TEST_CASE("Sys Functions")
 
   SECTION("health with supplied leader")
   {
-    auto response = sys.health(Vault::Path{"http://localhost:8200"});
+    auto response = sys.health(Vault::Url{"http://localhost:8200"});
 
     if (response) {
       auto keys = nlohmann::json::parse(response.value());

